index.html

```diff
@@ -0,0 +1,575 @@
+    mat4.multiply(g_mViewProjection, mProjection, mView);
+}
+
+function initRender()
+{
+    gl.viewport(0, 0, g_canvas.width, g_canvas.height);
+    gl.clearColor(0.0, 0.0, 0.0, 1.0);
+
+    g_pointSize = 4.0 * g_canvas.width * g_canvas.height / (1024 * 1024);
+
+    var vertexShader = loadShader(gl.VERTEX_SHADER, g_shaderStr[VS]);
+    var fragmentShader = loadShader(gl.FRAGMENT_SHADER, g_shaderStr[FS]);
+    g_programs[VS] = gl.createProgram();
+    gl.attachShader(g_programs[VS], vertexShader);
+    gl.attachShader(g_programs[VS], fragmentShader);
+
+    // Bind a_particlePos to attribute 0
+    // Bind a_particleVel to attribute 1
+    gl.bindAttribLocation(g_programs[VS], 0, "a_particlePos");
+    gl.bindAttribLocation(g_programs[VS], 1, "a_particleDen");
+
+    // Load the vertex data
+    gl.bindBuffer(gl.ARRAY_BUFFER, g_particleBuffers[0]);
+    gl.enableVertexAttribArray(0);
+    gl.vertexAttribPointer(0, 2, gl.FLOAT, gl.FALSE, 16, 0);
+    gl.vertexAttribDivisor(0, 1);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, g_particleBuffers[1]);
+    gl.enableVertexAttribArray(1);
+    gl.vertexAttribPointer(1, 1, gl.FLOAT, gl.FALSE, 16, 0);
+    gl.vertexAttribDivisor(1, 1);
+
+    // Link the program
+    gl.linkProgram(g_programs[VS]);
+    gl.useProgram(g_programs[VS]);
+
+    // Set uniforms
+    var mViewProjectionLoc = gl.getUniformLocation(g_programs[VS], "u_viewProjection");
+    gl.uniformMatrix4fv(mViewProjectionLoc, false, g_mViewProjection);
+    var fPointSizeLoc = gl.getUniformLocation(g_programs[VS], "u_fPointSize");
+    gl.uniform1f(fPointSizeLoc, g_pointSize);
+}
+
+function initCS() {
+    [CS_DEN_SIM, CS_DEN_SHA, CS_FOR_SIM, CS_FOR_SHA, CS_INT]
+    .forEach((v) => {
+        var cs = loadShader(gl.COMPUTE_SHADER, g_shaderStr[v]);
+        g_programs[v] = gl.createProgram();
+        gl.attachShader(g_programs[v], cs);
+        gl.linkProgram(g_programs[v]);
+    });
+}
+
+function destroyBuffers() {
+    gl.deleteBuffer(g_particleBuffers[0]);
+    gl.deleteBuffer(g_particleBuffers[1]);
+    gl.deleteBuffer(g_particleBuffers[2]);
+    gl.deleteBuffer(g_updateParams);
+}
+
+function computeDensityPass()
+{
+    gl.useProgram(g_currentDen);
+
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 1, g_particleBuffers[0]);
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 2, g_particleBuffers[1]);
+    gl.dispatchCompute(g_dispatchNum, 1, 1);
+}
+
+function computeForcePass()
+{
+    gl.useProgram(g_currentFor);
+
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 1, g_particleBuffers[0]);
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 2, g_particleBuffers[1]);
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 3, g_particleBuffers[2]);
+    gl.dispatchCompute(g_dispatchNum, 1, 1);
+}
+
+function computeIntegratePass()
+{
+    gl.useProgram(g_programs[CS_INT]);
+
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 1, g_particleBuffers[0]);
+    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 2, g_particleBuffers[0]);
```

Comment:
I just found that you are binding the same resource to different binding points. It's not supported by d3d backend. That's also why you can't get a correct result in d3d. Please change it like below:
```
    gl.bindBuffer(gl.COPY_READ_BUFFER, g_particleBuffers[0]);
    gl.bindBuffer(gl.COPY_WRITE_BUFFER, g_particleBuffers[3]);
    gl.copyBufferSubData(gl.COPY_READ_BUFFER, gl.COPY_WRITE_BUFFER, 0, 0, 16 * g_iNumParticles);
    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 1, g_particleBuffers[0]);
    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 2, g_particleBuffers[3]);
    gl.bindBufferBase(gl.SHADER_STORAGE_BUFFER, 3, g_particleBuffers[2]);
```
And remember to allocate a `g_particleBuffers[3]` in `initBuffers()`. I tested in my local side. It works well on d3d backend.